### ISSUE TITLE
Future of docker auto builds, continue to use dockerhub?

### ISSUE THREAD

**Issue by drwetter** [drwetter/testssl.sh#1939]
In the past whenever something was merged into 3.1dev or 3.0 a new container was build automatically containing the latest commit.

That is not anymore the case now, unless I pay for it:

![image](https://user-images.githubusercontent.com/8036727/127483649-12a6ff60-ec82-4110-8336-4e39997b7d7b.png)

Is there an alternative?

**Comment by jauderho** [comment]
You should be able to build using GitHub Actions and push to Docker Hub & GHCR. You can see an example here based on a release tag but will work the same way for based on push.

https://github.com/jauderho/dockerfiles/blob/main/.github/workflows/ansible.yml

**Comment by drwetter** [comment]
Thanks, @jauderho . What bugs me is that the credentials need to be store in clear text

**Comment by drwetter** [comment]
I am not even able to build manually with the webgui.

On the command line I have two problems

``` 
---> Running in 8564f3e503bd
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/main/x86_64/APKINDEX.tar.gz
ERROR: http://dl-cdn.alpinelinux.org/alpine/v3.12/main: DNS lookup error
WARNING: Ignoring http://dl-cdn.alpinelinux.org/alpine/v3.12/main: No such file or directory
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/community/x86_64/APKINDEX.tar.gz
ERROR: http://dl-cdn.alpinelinux.org/alpine/v3.12/community: DNS lookup error
WARNING: Ignoring http://dl-cdn.alpinelinux.org/alpine/v3.12/community: No such file or directory
```

On a server: ``docker login`` doesn't work as it requires X or some binaries to be installed which have X-dependencies

**Comment by drwetter** [comment]
The DNS error was due to a proxy. I believe the build command had a problem with that. Once I removed the proxy statement in config.json it worked.

So I just pushed the recent images manually until we find a proper solution

**Comment by jauderho** [comment]
> 
> 
> Thanks, @jauderho . What bugs me is that the credentials need to be store in clear text

I store the Docker credentials as a GitHub secret so I suppose it's not perfect but it's workable for now. The proper solution would be for Docker to provide proper API token capability. 

If this is a big concern, you can choose to publish only to GHCR which correctly uses just a token. 

---

The Dockerfile that you currently have assumes that there is a local copy available. To build using GH Actions, we will  have to checkout the latest commit upon push to initiate a build. 

No promises but I'll see if I have some time to throw something together soonish.

**Comment by drwetter** [comment]
> If this is a big concern, you can choose to publish only to GHCR which correctly uses just a token

--verbose, please?

**Comment by polarathene** [comment]
I don't know our workflow for publishing docker images too well, but we have [Github Actions setup for both DockerHub and GHCR if it's helpful to reference](https://github.com/docker-mailserver/docker-mailserver/blob/f33374023f94be9a6daf2093f3cf0bdbf8d1d7aa/.github/workflows/default_on_push.yml#L89-L99)

**Comment by drwetter** [comment]
Thanks!
Where is the dockerhub pw stored?

**Comment by polarathene** [comment]
> Where is the dockerhub pw stored?

I'm not sure, I don't have access to such. I only maintain docs CI and assist with PR reviews.

I've reached out to the other maintainers, so if one is able to assist they'll chime in here :)

**Comment by georglauterbach** [comment]
Chiming on from docker-mailserver/docker-mailserver. I was there when CI was set up, and you should have a look at

<https://github.com/docker-mailserver/docker-mailserver/blob/master/.github/workflows/default_on_push.yml>

The password is a secret one can set under "Repository -> Settings -> Secrets". For us, it looks like this:

![Screenshot 2021-08-04 at 16-39-51 Build software better, together](https://user-images.githubusercontent.com/44545919/128200958-72295a56-2654-48dc-af60-2dbdfe5af0aa.png)

**Comment by drwetter** [comment]
Thanks, @georglauterbach,  I found it.. 

As said above it's not a good solution because github owns my dockerhub password which is bad practice. Via token doesn't seem to exist.  

Maybe I should look into GHCR.

**Comment by casperklein** [comment]
> Via token doesn't seem to exist.

You can create Access-Tokens in Docker HUB. When logged in, goto https://hub.docker.com/settings/security

**Comment by jauderho** [comment]
Here's another useful reference: https://docs.docker.com/ci-cd/github-actions/

That guide should be sufficient to walk you through what needs to be done. But in summary, you can create a separate PAT different from your password that you can then store as a secret in GitHub to use to push into Docker Hub.

But it should be fine if you want to push just to GHCR. With my dockerfiles repo, I push to both and it works nicely as I do not need to previously prep a container config on Docker Hub prior to the first push.

**Comment by jauderho** [comment]
So I took some time tonight and threw something together that will build based on a release tag.

See commits (I forgot to fix the version in the first commit)

- https://github.com/jauderho/dockerfiles/commit/ad32f06ef17d5bb26381661890583953b6843765
- https://github.com/jauderho/dockerfiles/commit/b1a0c78cc2436abf01d19f6c410ed3f8ec3c61fe

You can see the resulting images at https://hub.docker.com/r/jauderho/testssl.sh/tags?page=1&ordering=last_updated and as an added bonus, there are images available for multiple architectures.

With a little backend cron job that I have running, it will build new images no longer than an hour after you tag a new release.

With a little bit more work, it should be possible to build a new image on every push.

**Comment by jauderho** [comment]
@drwetter 

I think I have things working, see 

- https://github.com/jauderho/sandbox/blob/main/docker/Dockerfile
- https://github.com/jauderho/sandbox/blob/main/.github/workflows/docker-3.0.yml
- https://github.com/jauderho/sandbox/blob/main/.github/workflows/docker-3.1dev.yml

This also takes advantage of the latest GH Action caches so builds/rebuilds are nice and fast. 

For next steps, if you can follow the link posted (https://docs.docker.com/ci-cd/github-actions/) to create the necessary PAT for Docker Hub, I can re-add the necessary yaml to be able to push to both Docker Hub and GHCR.

Once that is done, I can then submit a proper pull request to add it to your repo.

**Comment by drwetter** [comment]
Which right is the PAT supposed to have ? R+W+D?

**Comment by drwetter** [comment]
Oh, great. This sucks big time, @ dockerhub !

I can't have 2FA AND a PAT at the same time with a free account. WTF is this??

![image](https://user-images.githubusercontent.com/8036727/131126607-6eae5395-f420-478a-a892-c6d9e0056d4c.png)

**Comment by drwetter** [comment]
> Once that is done, I can then submit a proper pull request to add it to your repo

You can give it a try, just for testing. But actually I am not sure now whether I want this on the long run, see at least reasons above.

**Comment by jauderho** [comment]
@drwetter I ran into the same problem when I was testing this for testssl.sh as I had spun up my fork of testssl.sh and it was not letting me push to Docker Hub via a separate repo.

I still like Docker Hub because it shows the size of the images but yeah, maybe GHCR is the way forward as Docker dev in general seems to not be moving along anymore.

**Comment by drwetter** [comment]
A short heads up: probably I end up either using a manual work flow with ``docker login`` and then push the image or I might end up paying USD 60 to get automated build like I had before. There were people using the paypal button on https://testssl.sh so that's a way to spend the money. I leave this open as a reminder for a todo for me.

As helpful as the infrastructure is by some companies providing services for developing open source software is: this reminds me on how open source is depending on such a vendor based service. I hope at a certain point of time Microsoft doesn't change their ToS for github. Or it will be sold and the new owner charges money (see Solaris, Java and Sun --> Oracle)

**Comment by drwetter** [comment]
I just paid for a subscription, using money people spent for this project. Thank you!

That guarantees that the docker autobuilds take place again and less obstacles e.g. wrt the number of accounts are encountered. That seem to happen seemlessly, as soon as some coins were inserted the menu items reappeared.

A lucky and healthy year in 2022!

**Comment by jauderho** [comment]
@drwetter Do you want me to submit pull requests to add Docker login to the github actions? Or are you planning to build on Docker Hub?

**Comment by drwetter** [comment]
Thanks for asking. I wasn't really fond of that IIRC. I would say let's see whether the web hooks in dockerhub work as expected.

**Comment by drwetter** [comment]
FYI: The recent merge was applied to the image @ dockerhub automatically.  That happened purely with _Autobuild_. Not a webhook. 

![image](https://user-images.githubusercontent.com/8036727/147965712-303433e6-e1cb-4468-b4c3-37c7fd8b5b04.png)

So, everything's fine again
